### PR TITLE
Matching keys lengths and new metrics

### PR DESCRIPTION
**Pull request by IdoAmosIBM**
Added 2 new metrics: mathews correlation coef. and balanced accuracy.
Added a new functionality to default collate allowing to match the length of some keys in the batch dict to other keys - example: for encoder model, matching the length of the labels to the length of the input after crop padding is applied.

### CODE REVIEW

File: fuse/eval/metrics/metrics_common.py [hunk 1]
```diff
@@ -160,7 +160,10 @@ def collect(self, batch: Dict) -> None:
             batch_to_collect = {}
 
             for name, key in self._keys_to_collect.items():
-                value = batch[key]
+                try:
```
Comment (mosheraboh) [inline fuse/eval/metrics/metrics_common.py:163]: Was it just for debugging?
Or would you like to print a more informative message and then reraise the exception?

File: fuse/eval/metrics/classification/metrics_classification_common.py [hunk 2]
```diff
@@ -336,3 +337,81 @@ def __init__(self, pred: str, target: str, **kwargs: dict):
...
+            pred=pred,
+            target=target,
+            sample_weight=sample_weight,
+            metric_func=self.mcc_wrapper,
+            **kwargs,
+        )
+
+    def mcc_wrapper(
+        self,
+        pred: Optional[str] = None,
```
Comment (mosheraboh) [inline fuse/eval/metrics/classification/metrics_classification_common.py:368]: is it a string or a list of numpy array?
Are you expecting class predictions or scores after sofrmax?

File: fuse/data/utils/collates.py [hunk 3]
```diff
@@ -247,3 +255,22 @@ def crop_padding(
         cropped_sequences = [ids[:min_length] for ids in input_ids_list]
         batched_sequences = torch.stack(cropped_sequences, dim=0)
         return batched_sequences
+
+    @staticmethod
+    def match_length_to_target_key(
```
Comment (mosheraboh) [inline fuse/data/utils/collates.py:260]: Cropping?
What happens when the target_key is shorter then keys_to_match?

File: fuse/data/utils/collates.py [hunk 4]
```diff
@@ -43,6 +43,7 @@ def __init__(
         keep_keys: Sequence[str] = tuple(),
         raise_error_key_missing: bool = True,
         special_handlers_keys: Optional[Dict[str, Callable]] = None,
+        post_collate_special_handlers_keys: Optional[List[Callable]] = None,
```
Comment (mosheraboh) [inline fuse/data/utils/collates.py:46]: Why not to use special_handlers_keys instead?
Can you give an example of a use case?

**Review by mosheraboh — COMMENTED** [review]
Thanks Ido!
Few comments and questions inline.

File: fuse/data/utils/collates.py (hunk 3 above)
Comment (IdoAmosIBM) [reply fuse/data/utils/collates.py:260]: this will raise an error, maybe I'll change the name length to `crop_length_to_target_key`? The motivation is when the encoder and labels should have the same length and pad cropping was applied to the encoder input - so we want to match the labels.

File: fuse/data/utils/collates.py (hunk 4 above)
Comment (IdoAmosIBM) [reply fuse/data/utils/collates.py:46]: a use case is `match_target_key_to_length` from below, we can't use special handlers because it handles one element in the batch dict each time, we need to match one key to another after the special handler has been applied.

File: fuse/data/utils/collates.py (hunk 3 above)
Comment (mosheraboh) [reply fuse/data/utils/collates.py:260]: I assume you mean decoder input?
I asked Michal to solve it in a different way.
(labels and decoder should always have the same length).

File: fuse/data/utils/collates.py (hunk 4 above)
Comment (mosheraboh) [reply fuse/data/utils/collates.py:46]: Why is it a list and not a single callable?

File: fuse/data/utils/collates.py (hunk 3 above)
Comment (IdoAmosIBM) [reply fuse/data/utils/collates.py:260]: sorry, I do mean encoder, this is for the case of an encoder-only model

File: fuse/eval/metrics/classification/metrics_classification_common.py [hunk 5]
```diff
@@ -336,3 +337,81 @@ def __init__(self, pred: str, target: str, **kwargs: dict):
...
+            pred=pred,
+            target=target,
+            sample_weight=sample_weight,
+            metric_func=self.balanced_acc_wrapper,
+            **kwargs,
+        )
+
+    def balanced_acc_wrapper(
+        self,
+        pred: Union[List, np.ndarray],
```
Comment (mosheraboh) [inline fuse/eval/metrics/classification/metrics_classification_common.py:407]: It's still not clear if you expect here sofrmax scores (we call pred) or class predictions (we call cls_pred).

**Review by mosheraboh — DISMISSED** [review]
Thanks Ido!
I still have few comments.
In order not to block you, I'm ok with merging and potentially make some modifications in a new PR.

File: fuse/data/utils/collates.py (hunk 4 above)
Comment (IdoAmosIBM) [reply fuse/data/utils/collates.py:46]: to allow multiple use cases, e.g. aligning a set of keys to key A and a different set of keys to key B.

File: fuse/eval/metrics/classification/metrics_classification_common.py (hunk 5 above)
Comment (IdoAmosIBM) [reply fuse/eval/metrics/classification/metrics_classification_common.py:407]: oh I understand, I expect class predictions I used the same names as in MetricAccuracy. Do you want me to change to cls_pred?

**Review by SagiPolaczek — APPROVED** [review]
LGTM! 
Thanks!
